#6098: Allowing to pass request headers from a spider parser to the spider task

vladidx opened the pull request:
- Introduction of a new class 'SpiderResourceFound' storing resource
details used within SpiderParser, SpiderParserListener,
SpiderController, SpiderTask, and SpiderParserTestUtils
- Simplified canonicalization of visited resources (incl. supplied
headers) using one hash set
- Introduction of new unit tests in class
'SpiderParserAdditionalHeadersUnitTest' which create a custom test
spider parser + controller (incl. spider) and cover header-related
handling in the spider controller

Signed-off-by: Vladislav Dexheimer <vladislav.dexheimer@sap.com>

vladidx commented:
@thc202 I'm very sorry. I unintentionally closed the previous pull request #6104 due to a hard reset and I'm not able to reopen it again.

I hope we can continue the discussion here for #6098. I refactored parts of the spider resource handling as you mentioned and I hope it suits the expectations. As discussed, this breaks the binary compatibility in the SpiderParserListener class.

vladidx commented:
@thc202: any chance to review and include this into the upcoming ZAP version 2.10? I would like to make use of this functionality as part of an own extension.

kingthorin commented:
Considering the build failure and conflicts that doesn't seem practical.

thc202 commented:
The release will be done today (that's the plan at least).

vladidx commented:
Got it, thanks. The build failure results from the expected binary incompatibility as discussed in #6104.
I will try to resolve the conflicts. Maybe it's then possible to schedule this PR after the release for 2.11?

thc202 commented:
Yes, #6098 has been added to the 2.11 milestone.

thc202 commented:
You'll have to suppress the incompatible changes in `zap/zap.gradle.kts`, under the `japicmp` task. You can run that task to have the report and know what needs to be suppressed.

I'd suggest rebase first on the `main` branch though.

vladidx commented:
The PR is now rebased on branch `main` and I've added the method exclusions to the `japicmp` task.
All checks are green now, I hope this is sufficient.

thc202 commented on zap/src/main/java/org/zaproxy/zap/spider/parser/SpiderParser.java:
```diff
@@ -40,7 +42,7 @@
      * The Constant log.
      *
-     * @deprecated (2.10.0) Use {@link #getLogger()} instead.
+     * @deprecated (TODO add version) Use {@link #getLogger()} instead.
```
Should be reverted.

vladidx replied:
Resolved via last commit

thc202 commented on zap/src/main/java/org/zaproxy/zap/spider/parser/SpiderParser.java:
```diff
@@ -52,7 +54,7 @@
      *
      * @return the logger, never {@code null}.
-     * @since 2.10.0
+     * @since TODO add version
```
Should be reverted.

vladidx replied:
Resolved via last commit

thc202 commented on zap/src/main/java/org/zaproxy/zap/spider/parser/SpiderParser.java:
```diff
@@ -76,6 +78,18 @@ public void removeSpiderParserListener(SpiderParserListener listener) {
+     * @param resourceFound the http message containing the response.
+     * @since 2.11.0
+     */
+    protected void notifyListenersResourceFound(SpiderResourceFound resourceFound) {
```
This should be the only method being added, older methods should be deprecated.

vladidx replied:
Resolved via last commit, previous notify... methods are now deprecated

thc202 commented on zap/src/main/java/org/zaproxy/zap/spider/parser/SpiderResourceFound.java:
```diff
@@ -0,0 +1,178 @@
+ *
+ * @since 2.11.0
+ */
+public class SpiderResourceFound {
```
Better provide a builder, can be immutable.

vladidx replied:
Resolved via last commit, SpiderResourceFound now offers a builder returning immutable instances

thc202 commented on zap/src/main/java/org/zaproxy/zap/spider/parser/SpiderResourceFound.java:
```diff
@@ -0,0 +1,178 @@
+    /** Defines resource as useful or not useful in the fetching process. */
+    private boolean shouldIgnore = false;
+    /** Additional request headers to be passed for the resource. */
+    private List<HttpHeaderField> requestHeaders = new ArrayList<>();
```
The default should be `Collections.emptyList()`, otherwise it will create a lot of garbage.

vladidx replied:
Resolved via last commit, now part of the builder

thc202 commented on zap/src/main/java/org/zaproxy/zap/spider/SpiderTask.java:
```diff
@@ -76,73 +78,20 @@
      */
-    public SpiderTask(
-            Spider parent, URI sourceUri, URI uri, int depth, String method, String requestBody) {
+    public SpiderTask(Spider parent, SpiderResourceFound resoureFound, URI uri) {
```
Typo in resource.

vladidx replied:
Corrected via last commit

thc202 commented on zap/src/main/java/org/zaproxy/zap/spider/parser/SpiderResourceFound.java:
```diff
@@ -0,0 +1,178 @@
+ * @since 2.11.0
+ */
+public class SpiderResourceFound {
+    /** Original response message. */
```
`The message where the resource was found.`

vladidx replied:
Corrected via last commit

thc202 commented on zap/src/main/java/org/zaproxy/zap/spider/parser/SpiderResourceFound.java:
```diff
@@ -0,0 +1,178 @@
+ */
+public class SpiderResourceFound {
+    /** Original response message. */
+    private HttpMessage responseMessage;
```
`message`.

vladidx replied:
Corrected via last commit

thc202 commented on zap/src/main/java/org/zaproxy/zap/spider/SpiderTask.java:
```diff
@@ -153,14 +102,29 @@ public SpiderTask(
+                }
+            }
+            if (resoureFound.getResponseMessage() != null
+                    && resoureFound.getResponseMessage().getRequestHeader() != null
```
The request header is never null.

vladidx replied:
Corrected via last commit

thc202 commented on zap/src/main/java/org/zaproxy/zap/spider/SpiderController.java:
```diff
@@ -251,46 +248,71 @@ public void init() {
+    private String buildCanonicalResourceIdentifier(URI uri, SpiderResourceFound resourceFound)
+            throws URIException {
+        StringBuilder identifierBuilder = new StringBuilder(50);
+        if (resourceFound != null) {
```
Is this needed?

vladidx replied:
Removed check in last commit as not needed

thc202 commented on zap/src/main/java/org/zaproxy/zap/spider/SpiderController.java:
```diff
@@ -251,46 +248,71 @@ public void init() {
+    @Override
+    public void resourceFound(SpiderResourceFound resourceFound) {
+        if (resourceFound == null
+                || resourceFound.getUri() == null
```
Better do these validations when the resource is created, here it should just be used.

vladidx replied:
Resolved via last commit, validations are now done as part of the builder

thc202 commented on zap/src/main/java/org/zaproxy/zap/spider/SpiderController.java:
```diff
@@ -251,46 +248,71 @@ public void init() {
             return;
         }
 
+        log.debug("New " + resourceFound.getMethod() + "resource found: " + resourceFound.getUri());
```
`log.debug("New {} resource found: {}", resourceFound.getMethod(),resourceFound.getUri())`

vladidx replied:
Corrected via last commit

thc202 commented on zap/src/main/java/org/zaproxy/zap/spider/SpiderController.java:
```diff
@@ -299,107 +321,51 @@ public void resourceURIFound(
-            }
+    private String getCanonicalHeadersString(List<HttpHeaderField> headers) {
+        if (headers != null && !headers.isEmpty()) {
+            headers.removeIf(
```
Better do these validations when the resource is created, here it should just be used.

vladidx replied:
Resolved via last commit, validations are now done as part of the builder

thc202 commented:
I still need to go through the tests and the `SpiderController` more closely.

vladidx commented:
Thanks for the review. I'll provide the requested changes within the next two days.

vladidx commented:
The requested changes are now implemented via the last commit. Please let me know if anything is missing or requires adjustments.

thc202 commented:
The test `SpiderParserAdditionalHeadersUnitTest` needs to be simplified/split, it's not an unit test right now (more integration/functional). It would be better to test the parts individually as a start (I didn't check what can be done yet though).

vladidx commented:
Thanks for the update. I will try to split the unit test into individual parts.

vladidx commented:
@thc202 The former unit test is now split into two separate ones:
- SpiderParserGenericWithBuilderUnitTest: Aims to test the SpiderResourceFound builder as part of a generic spider parser
- SpiderControllerResourcesFoundUnitTest: Aims to test a generic spider controller when it comes to the resource canonicalization and task submission

Please let me know if this separation is OK.

thc202 commented:
Rebased to pick #6782 (also squashed).

thc202 commented:
Rebased to pick #6794 and add additional tweaks on top (split into their own commits for clarity, will squash once reviewed).

kingthorin reviewed (approved):
Thanks!

kingthorin commented:
There are a few files that seem new but have 2020 in their license block, this is likely because of the date of the original PR. Should that be tweaked or does it not really matter?

thc202 commented:
Will update them and fix up.

thc202 commented:
Done.

kingthorin reviewed (approved):
TY!

thc202 commented:
Thank you!
